dali/operators/decoder/numpy.cc

```diff
@@ -0,0 +1,157 @@
+  for (int sampleIdx = 0; sampleIdx < input.num_samples(); ++sampleIdx) {
+    const auto &header = headers_[sampleIdx];
+    if (header.shape.sample_dim() != output_shape.sample_dim()) {
+      DALI_FAIL(
+          make_string("All samples in the batch must have the same number of dimensions, "
+                      "but got differing sample dimensions: ",
+                      output_shape.sample_dim(), " and ", header.shape.sample_dim()));
+    }
+    if (header.fortran_order) {
+      // Fortran order means the shape is transposed, we need to reverse
+      auto transposed_shape = TensorShape<-1>::empty_shape(output_shape.sample_dim());
+      std::reverse_copy(header.shape.begin(), header.shape.end(), transposed_shape.begin());
+      output_shape.set_tensor_shape(sampleIdx, transposed_shape);
+    } else {
+      output_shape.set_tensor_shape(sampleIdx, header.shape);
+    }
+  }
+  output_desc[0].shape = output_shape;
+
+  return true;
+}
+
+void RunDecoding(SampleView<CPUBackend> outputSample, ConstSampleView<CPUBackend> inputView,
+                 const numpy::HeaderData &header) {
+  Tensor<CPUBackend> transposeBuffer;
+  if (header.fortran_order) {
+    transposeBuffer.Resize(outputSample.shape(), inputView.type());
+    auto transposeBufferView = SampleView<CPUBackend>(
+        transposeBuffer.raw_mutable_data(), transposeBuffer.shape(), transposeBuffer.type());
+    numpy::FromFortranOrder(transposeBufferView, inputView);
+    inputView = ConstSampleView<CPUBackend>(transposeBuffer.raw_data(), transposeBuffer.shape(),
+                                            transposeBuffer.type());
+  }
+
+  if (outputSample.type() != inputView.type()) {
+    // If the types do not match, we need to convert the data
+    TYPE_SWITCH(outputSample.type(), type2id, OType, NUMPY_ALLOWED_TYPES, (
+      TYPE_SWITCH(inputView.type(), type2id, IType, NUMPY_ALLOWED_TYPES, (
+        std::transform(inputView.data<IType>(),
+          inputView.data<IType>() + volume(inputView.shape()),
+          outputSample.mutable_data<OType>(), ConvertSat<OType, IType>);),
+      DALI_FAIL(make_string("Unsupported input type: ", inputView.type())););),
+    DALI_FAIL(make_string("Unsupported output type: ", outputSample.type())););
+  } else {
+    // If the types match, we can just copy the data
+    auto *out_ptr = outputSample.raw_mutable_data();
+    const auto *in_ptr = inputView.raw_data();
+    std::memcpy(out_ptr, in_ptr, header.nbytes());
+  }
+}
+
+
+void NumpyDecoder::RunImpl(Workspace &ws) {
+  const auto &input = ws.Input<CPUBackend>(0);
+  auto &output = ws.Output<CPUBackend>(0);
+  auto &tPool = ws.GetThreadPool();
+
+  for (int sampleIdx = 0; sampleIdx < input.num_samples(); ++sampleIdx) {
+    tPool.AddWork([&, sampleIdx](int) {
+      const numpy::HeaderData &header = headers_[sampleIdx];
+      ConstSampleView<CPUBackend> inputView(input.raw_tensor(sampleIdx) + header.data_offset,
```

Comment:
Clang complains about this line
```
#14 154.3 /opt/dali/dali/operators/decoder/numpy.cc:126:73: error: arithmetic on a pointer to void
#14 154.3   126 |       ConstSampleView<CPUBackend> inputView(input.raw_tensor(sampleIdx) + header.data_offset,
#14 154.3       |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```
indeed, the raw_tensor call returns `void*`. Adding a static_cast should make clang happy, something like:
```suggestion
      ConstSampleView<CPUBackend> inputView(static_cast<const uint8_t *>(input.raw_tensor(sampleIdx)) + header.data_offset,
```